core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleQueue.java

```diff
@@ -53,7 +53,7 @@ private void addMatch(VolcanoRuleMatch match, Deque<VolcanoRuleMatch> queue) {
       return;
     }
 
-    if (!planner.isSubstituteRule(match)) {
+    if (planner.isSubstituteRule(match)) {
       queue.addFirst(match);
```

Comment:
Here is the main change.

Reply:
Why it is still works before this patch ? Does that mean the place in the rule queue has no effect for SubstituteRule ?

Reply:
The order of the rule does not have an impact on the final plan. But it has an impact on the efficiency of the optimizer.

Reply:
How much gains for this patch ?